### PR TITLE
Add Jenkinsfile for nightly builds

### PR DESCRIPTION
**Pull request by dalg24**
At the moment adding only one CUDA build against Kokkos develop branch

NOTE: Boost from the system package manager is version 1.71 which yields error with NVCC when enabling the tests .

### CODE REVIEW

File: .jenkins/nightly.groovy [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                        }
+                    }
+                    environment {
+                        CTEST_OPTIONS = '--no-compress-output -T Test'
+                        CMAKE_OPTIONS = '-D CMAKE_BUILD_TYPE=Release -D CMAKE_CXX_STANDARD=17 -D CMAKE_CXX_EXTENSIONS=OFF'
+                    }
+                    steps {
+                        sh 'apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y git cmake libboost-program-options-dev libboost-test-dev libbenchmark-dev'
+                        sh 'rm -rf source* build* install*'
+                        sh 'git clone https://github.com/kokkos/kokkos.git --branch develop --depth 1 source-kokkos'
```
Comment (aprokop) [inline .jenkins/nightly.groovy:22]:
```suggestion
                        sh 'cd source-kokkos && rev-parse --short HEAD && cd -'
```

File: .jenkins/nightly.groovy [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                stage('CUDA-11.4.2') {
+                    agent {
+                        docker {
+                            image 'nvidia/cuda:11.4.2-devel-ubuntu20.04'
+                            label 'NVIDIA_Tesla_V100-PCIE-32GB && nvidia-docker'
+                        }
+                    }
+                    environment {
+                        CTEST_OPTIONS = '--no-compress-output -T Test'
+                        CMAKE_OPTIONS = '-D CMAKE_BUILD_TYPE=Release -D CMAKE_CXX_STANDARD=17 -D CMAKE_CXX_EXTENSIONS=OFF'
```
Comment (aprokop) [inline .jenkins/nightly.groovy:17]: Do we care what CMake build type to use here? I can't think of any benefits of one over the other.

**Review by aprokop — APPROVED** [review]
I think this is a really good idea.

File: .jenkins/nightly.groovy (hunk 2 above)
Comment (dalg24) [reply .jenkins/nightly.groovy:17]: I just wanted consistent settings between the configurations.  The actual build type does not matter.

**Comment by dalg24** [comment]
Failures are unrelated (fetnat03 running out of space)
I should have skipped the CI anyway

File: .jenkins/nightly.groovy [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            parallel {
+                stage('CUDA-11.4.2') {
+                    agent {
+                        docker {
+                            image 'nvidia/cuda:11.4.2-devel-ubuntu20.04'
+                            label 'NVIDIA_Tesla_V100-PCIE-32GB && nvidia-docker'
+                        }
+                    }
+                    environment {
+                        CTEST_OPTIONS = '--no-compress-output -T Test'
```
Comment (masterleinad) [inline .jenkins/nightly.groovy:16]: What about adding the same timeout we have for pull request testing?
```suggestion
                        CTEST_OPTIONS = '--timeout 180 --no-compress-output -T Test'
```

**Review by masterleinad — APPROVED** [review]
Looks good to me.

File: .jenkins/nightly.groovy [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                        sh 'cmake --install build-kokkos'
+                        sh 'cmake -B build-arborx -D CMAKE_INSTALL_PREFIX=$PWD/install-arborx $CMAKE_OPTIONS -D Kokkos_ROOT=$PWD/install-kokkos'
+                        sh 'cmake --build build-arborx --parallel 8'
+                        sh 'cmake --install build-arborx'
+                        sh 'cmake -S examples -B build-examples -D ArborX_ROOT=$PWD/install-arborx -D Kokkos_ROOT=$PWD/install-kokkos $CMAKE_OPTIONS'
+                        sh 'cmake --build build-examples --parallel 8'
+                        dir('build-examples') {
+                            sh 'ctest $CTEST_OPTIONS'
+                        }
+                        sh 'cmake -S examples -B build-benchmarks -D ArborX_ROOT=$PWD/install-arborx -D Kokkos_ROOT=$PWD/install-kokkos $CMAKE_OPTIONS'
```
Comment (dalg24) [inline .jenkins/nightly.groovy:37]: Upppss `-S examples -> -S benchmarks`...

Comment (dalg24) [reply .jenkins/nightly.groovy:37]: And turns out it does not work...

**Pull request merged**